lib/PuppeteerSharp/WaitTask.cs

```diff
@@ -183,84 +87,127 @@ public void Dispose()
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
+                }
+                else
                 {
-                    await success.DisposeAsync().ConfigureAwait(false);
+                    _poller = await _isolatedWorld.EvaluateFunctionHandleAsync(
+                            @"
+                            ({MutationPoller, createFunction}, root, fn, ...args) => {
+                                const fun = createFunction(fn);
+                                return new MutationPoller(() => {
+                                return fun(...args);
+                                }, root || document);
+                            }",
+                            new object[]
+                            {
+                            await _isolatedWorld.GetPuppeteerUtilAsync().ConfigureAwait(false),
+                            _root,
+                            _fn,
+                            }.Concat(_args).ToArray()).ConfigureAwait(false);
                 }
 
-                return;
-            }
+                await _poller.EvaluateFunctionAsync("poller => poller.start()").ConfigureAwait(false);
 
-            if (exception == null &&
-                await _isolatedWorld.EvaluateFunctionAsync<bool>("s => !s", success)
-                    .ContinueWith(
-                        task => task.IsFaulted || task.Result,
-                        TaskScheduler.Default)
-                    .ConfigureAwait(false))
+                var success = await _poller.EvaluateFunctionHandleAsync("poller => poller.result()").ConfigureAwait(false);
+                _result.TrySetResult(success);
+                await TerminateAsync().ConfigureAwait(false);
+            }
+            catch (Exception ex)
             {
-                if (success != null)
+                var exception = GetBadException(ex);
+                if (exception != null)
                 {
-                    await success.DisposeAsync().ConfigureAwait(false);
+                    await TerminateAsync(exception).ConfigureAwait(false);
                 }
-
-                return;
             }
+        }
 
-            if (exception?.Message.Contains("Execution context was destroyed") == true)
+        internal async Task TerminateAsync(Exception exception = null)
+        {
+            // The timeout timer might call this method on cleanup
+            if (_terminated)
             {
-                _ = Rerun();
                 return;
             }
 
-            if (exception?.Message.Contains("Cannot find context with specified id") == true)
-            {
-                return;
-            }
+            _terminated = true;
+            _isolatedWorld.TaskManager.Delete(this);
+            Cleanup(); // This matches the clearTimeout upstream
```

Comment:
`_isolatedWorld.TaskManager.Delete(this);` is also called inside `Cleanup`

Reply:
Good catch. Removing it from `Cleanup`.